### PR TITLE
feat(ST-86): create classroom endpoint 

### PR DESCRIPTION
**Pull request by ghost**
## Description of Change
- Added Classroom module, controller, service.
- Updated the classroom repository file.
- Added dto for CreateClassroom and also for ClassroomResponse.
- Added protected route for the create classroom endpoint using jwt auth guard.
- Added the response type in the auth.interface.ts file.

## Associated Ticket Link(s)

- https://trello.com/c/VM5HQrD5

## Screenshots / Screen Recordings
N/A

### CODE REVIEW

File: src/classrooms/classrooms.controller.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { Body, Controller, Post, UseGuards, Req } from "@nestjs/common";
+
+import { AuthenticatedRequest } from "@/auth/auth.interfaces";
+import { JwtAuthGuard } from "@/auth/guards/jwt-auth.guard";
+
+import { ClassroomResponseDto, CreateClassroomDto } from "./classrooms.dtos";
+import { ClassroomsService } from "./classrooms.service";
+
+@Controller("classrooms")
+@UseGuards(JwtAuthGuard)
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts]: Let's move the guard above the controller decorator

File: src/classrooms/classrooms.service.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      { id: authenticatedUserId },
+      { populate: ["teacher"] },
+    );
+    if (!user) {
+      throw new NotFoundException("User not found");
+    }
+
+    if (!user.teacher) {
+      throw new ForbiddenException("Only teachers can create classrooms");
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: Should use roles decorator at the controller to achieve this.

Can find this in the template for createUser in users.controller.

File: src/classrooms/classrooms.service.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    const classroom = new Classroom();
+    classroom.title = title;
+    classroom.subject = subject;
+    classroom.classTime = new Date(classTime);
+    classroom.days = days;
+    classroom.teacher = user.teacher;
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: Let's not create new records like this. Should simply call the classroom repository createOne and pass the classroom dto.

File: src/classrooms/classrooms.repository.ts [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
 
 import { Classroom } from "@/common/entities/classrooms.entity";
 
 @Injectable()
-export class ClassroomsRepository extends EntityRepository<Classroom> {}
+export class ClassroomsRepository extends EntityRepository<Classroom> {
+  async createAndSave(classroom: Classroom): Promise<Classroom> {
+    await this.em.persistAndFlush(classroom);
+    return classroom;
+  }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.repository.ts]: Don't need this method. Use createOne and persistAndFlush in the service layer.

File: src/classrooms/classrooms.controller.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+export class ClassroomsController {
+  constructor(private readonly classroomsService: ClassroomsService) {}
+
+  @Post()
+  createClassroom(
+    @Req() req: AuthenticatedRequest,
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts]: If you want the current user id, you can use the @CurrentUser decorator. You can find it in the template

File: src/classrooms/classrooms.controller.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import { ClassroomResponseDto, CreateClassroomDto } from "./classrooms.dtos";
+import { ClassroomsService } from "./classrooms.service";
+
+@UseGuards(JwtAuthGuard, RolesGuard)
+@Controller("classrooms")
+export class ClassroomsController {
+  constructor(private readonly classroomsService: ClassroomsService) {}
+
+  @Post()
+  @Roles(EUserType.TEACHER)
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts:18]: Nice, guard implementation looks good. But please be mindful that the roles guard at the Controller level might interfere with other endpoints.

File: src/classrooms/classrooms.controller.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { Body, Controller, Post, UseGuards } from "@nestjs/common";
+
+import { CurrentUser } from "@/auth/decorators/current-user.decorator";
+import { Roles } from "@/auth/decorators/roles.decorator";
+import { JwtAuthGuard } from "@/auth/guards/jwt-auth.guard";
+import { RolesGuard } from "@/auth/guards/roles.guard"; // Import the new RolesGuard
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts:6]: Should remove comments

File: src/classrooms/classrooms.service.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    return new ClassroomResponseDto({
+      id: classroom.id,
+      title: classroom.title,
+      subject: classroom.subject,
+      classTime: classroom.classTime,
+      days: classroom.days,
+      userId: user.id,
+    });
+  }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts:49]: Please remove this when serializer is implementated

**Review by ishmam-iqbal-sazim — APPROVED** [review]
Left some comments

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**